File: bench/python/ops/bench_warpaffine.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""CV-CUDA WarpAffine operator benchmark - Python equivalent of BenchWarpAffine.cpp"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import cupy as cp  # noqa: E402
import cvcuda  # noqa: E402
from python_bench_utils import (  # noqa: E402
    get_input_kind,
    parse_shape,
    get_dtype,
    get_dtype_size,
    get_num_channels,
    get_format_from_dtype,
    get_border_type,
    get_interpolation_type,
    create_tensor,
    create_stream_cache,
    create_image_batch_varshape,
    run_benchmark,
)


def warpaffine(state):
    """WarpAffine operator benchmark matching C++ BenchWarpAffine.cpp"""

    shape = parse_shape(state.get_string("shape"))
    dtype_str = state.get_string("InOutDataType")
    dtype = get_dtype(dtype_str)
    try:
        layout = state.get_string("layout")
    except Exception:
        layout = "NHWC"
    border = get_border_type(state.get_string("border"))
    interp = get_interpolation_type(state.get_string("interpolation"))
    inverse_map = state.get_string("inverseMap") == "Y"
    input_kind = get_input_kind(state.get_string("inputKind"))
    device_id = state.get_device()

    N, H, W = shape
    nc = get_num_channels(dtype_str)

    if layout not in ("NHWC", "NCHW", "NCHW_FAKE"):
        state.skip(
            "WarpAffine benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts"
        )
        return None

    is_planar = layout == "NCHW"
    is_fake_planar = layout == "NCHW_FAKE"
    if is_fake_planar and input_kind == "VarShape":
        state.skip("Fake-planar (NCHW_FAKE) WarpAffine benchmark is tensor-only")
        return None

    flags = interp
    if inverse_map:
        flags = flags | cvcuda.Interp.WARP_INVERSE_MAP

    # WarpAffine preserves size; the 6-float matrix is also read each launch.
    dtype_size = get_dtype_size(dtype_str)
    bytes_ = N * H * W * dtype_size
    if is_fake_planar:
        state.add_global_memory_reads(3 * bytes_ + 6 * 4)
        state.add_global_memory_writes(3 * bytes_)
    else:
        state.add_global_memory_reads(bytes_ + 6 * 4)
        state.add_global_memory_writes(bytes_)

    get_stream = create_stream_cache()

    if is_fake_planar:
        # Tensor-only "fake planar": planar->interleaved->warp->interleaved->planar, all timed, as the
        # comparison baseline for the native planar (NCHW) path.
        src = create_tensor(
            (N, nc, H, W), dtype, device_id, layout="NCHW", fill_mode="checkerboard"
        )
        inter_src = create_tensor(
            (N, H, W, nc), dtype, device_id, layout="NHWC", fill_mode=0
        )
        inter_dst = create_tensor(
            (N, H, W, nc), dtype, device_id, layout="NHWC", fill_mode=0
        )
        dst = create_tensor((N, nc, H, W), dtype, device_id, layout="NCHW", fill_mode=0)

        xform = [[2.0, 2.0, 0.0], [3.0, 1.0, 0.0]]

        def run_fake(launch):
            stream = get_stream(launch)
            cvcuda.reformat_into(inter_src, src, stream=stream)
            cvcuda.warp_affine_into(
                inter_dst,
                inter_src,
                xform,
                flags=flags,
                border_mode=border,
                border_value=(0, 0, 0, 0),
                stream=stream,
            )
            cvcuda.reformat_into(dst, inter_dst, stream=stream)

        return run_fake

    if input_kind == "Tensor":
        tshape = (N, nc, H, W) if is_planar else (N, H, W, nc)
        src = create_tensor(
            tshape, dtype, device_id, layout=layout, fill_mode="checkerboard"
        )
        dst = create_tensor(tshape, dtype, device_id, layout=layout, fill_mode=0)

        xform = [[2.0, 2.0, 0.0], [3.0, 1.0, 0.0]]

        def run(launch):
            stream = get_stream(launch)
            cvcuda.warp_affine_into(
                dst,
                src,
                xform,
                flags=flags,
                border_mode=border,
                border_value=(0, 0, 0, 0),
                stream=stream,
            )

        return run

    else:
        img_format = get_format_from_dtype(dtype_str, nc, planar=is_planar)

        src = create_image_batch_varshape(
            (N, H, W, nc),
            0,
            img_format,
            dtype,
            device_id,
            fill_mode="checkerboard",
        )
        dst = create_image_batch_varshape(
            (N, H, W, nc), 0, img_format, dtype, device_id, fill_mode=0
        )

        xform_data = cp.array([[2.0, 2.0, 0.0, 3.0, 1.0, 0.0]] * N, dtype=cp.float32)
        xform_tensor = cvcuda.as_tensor(xform_data, "NW")

        def run(launch):
            stream = get_stream(launch)
            cvcuda.warp_affine_into(
                dst,
                src,
                xform_tensor,
                flags=flags,
                border_mode=border,
                border_value=(0, 0, 0, 0),
                stream=stream,
            )

        return run


if __name__ == "__main__":
    run_benchmark("warpaffine", warpaffine)

File: bench/cpp/ops/BenchWarpAffine.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../CppBenchUtils.hpp"
#include "ops/generated/BenchWarpAffineConfig.hpp"

#include <cvcuda/OpReformat.hpp>
#include <cvcuda/OpWarpAffine.hpp>

#include <nvbench/nvbench.cuh>

template<typename T>
inline void warpaffine(nvbench::state &state, nvbench::type_list<T>)
try
{
    int3                        shape     = benchutils::GetShape<3, int3>(state.get_string("shape"));
    auto                        layout    = benchutils::GetStringParam(state, "layout", "NHWC");
    const benchutils::InputKind inputKind = benchutils::GetInputKind(state.get_string("inputKind"));

    NVCVBorderType        borderType = benchutils::GetBorderType(state.get_string("border"));
    NVCVInterpolationType interpType = benchutils::GetInterpolationType(state.get_string("interpolation"));

    int flags = interpType | ((state.get_string("inverseMap") == "Y") ? NVCV_WARP_INVERSE_MAP : 0);

    float4 borderValue{0, 0, 0, 0};

    NVCVAffineTransform transMatrix{2.f, 2.f, 0.f, 3.f, 1.f, 0.f};

    const bool isPlanar     = layout == "NCHW";
    const bool isFakePlanar = layout == "NCHW_FAKE";
    if (layout != "NHWC" && layout != "NCHW" && layout != "NCHW_FAKE")
    {
        state.skip("WarpAffine benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts");
        return;
    }
    // NCHW_FAKE ("fake planar") is a tensor-only comparison path: planar data is reformatted to
    // interleaved, warped with the interleaved kernel, and reformatted back to planar — all timed
    // together — so the native planar path (NCHW) can be shown to be faster than this naive
    // convert→warp→convert pipeline.
    if (isFakePlanar && inputKind == benchutils::InputKind::VarShape)
    {
        state.skip("Fake-planar (NCHW_FAKE) WarpAffine benchmark is tensor-only");
        return;
    }

    using BT = typename nvcv::cuda::BaseType<T>;
    int ch   = nvcv::cuda::NumElements<T>;

    // WarpAffine preserves size, so src and dst hold the same number of bytes.
    const long bytes = static_cast<long>(shape.x) * shape.y * shape.z * sizeof(T);
    if (isFakePlanar)
    {
        state.add_global_memory_reads(3 * bytes + 6 * sizeof(float));
        state.add_global_memory_writes(3 * bytes);
    }
    else
    {
        state.add_global_memory_reads(bytes + 6 * sizeof(float));
        state.add_global_memory_writes(bytes);
    }

    cvcuda::WarpAffine op(shape.x);

    // clang-format off

    if (isFakePlanar) // tensor-only: planar→interleaved→warp→interleaved→planar
    {
        nvcv::Tensor src     ({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>());
        nvcv::Tensor interSrc({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor interDst({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor dst     ({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>());

        benchutils::FillTensor<BT>(src, benchutils::CheckerboardValues<BT>());

        cvcuda::Reformat reformatOp;

        benchutils::warmup_and_exec(state, BENCH_WARPAFFINE_WARMUP_ITERATIONS,
            [&op, &reformatOp, &src, &interSrc, &interDst, &dst, &transMatrix, &flags, &borderType, &borderValue](cudaStream_t s) {
                reformatOp(s, src, interSrc);                                       // NCHW → NHWC
                op(s, interSrc, interDst, transMatrix, flags, borderType, borderValue); // interleaved warp
                reformatOp(s, interDst, dst);                                       // NHWC → NCHW
            });
    }
    else if (inputKind == benchutils::InputKind::Tensor)
    {
        nvcv::Tensor src = isPlanar
                             ? nvcv::Tensor({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>())
                             : nvcv::Tensor({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor dst = isPlanar
                             ? nvcv::Tensor({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>())
                             : nvcv::Tensor({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());

        benchutils::FillTensor<BT>(src, benchutils::CheckerboardValues<BT>());

        benchutils::warmup_and_exec(state, BENCH_WARPAFFINE_WARMUP_ITERATIONS,
            [&op, &src, &dst, &transMatrix, &flags, &borderType, &borderValue](cudaStream_t s) {
                op(s, src, dst, transMatrix, flags, borderType, borderValue);
            });
    }
    else // zero and positive var shape means use ImageBatchVarShape
    {
        nvcv::ImageBatchVarShape src(shape.x);
        nvcv::ImageBatchVarShape dst(shape.x);

        if (isPlanar)
        {
            benchutils::FillPlanarImageBatch<T>(src, long2{shape.z, shape.y}, long2{0, 0});
            benchutils::FillPlanarImageBatch<T>(dst, long2{shape.z, shape.y}, long2{0, 0});
        }
        else
        {
            benchutils::FillImageBatch<T>(src, long2{shape.z, shape.y}, long2{0, 0},
                                          benchutils::CheckerboardValues<T>());
            // Use FillImageBatchLike to ensure dst has same per-sample shapes as src
            benchutils::FillImageBatchLike<T>(dst, src, [](const long4_16a &) { return T{0}; });
        }

        nvcv::Tensor transMatrixTensor({{shape.x, 6}, "NW"}, nvcv::TYPE_F32);

        benchutils::FillTensor<float>(transMatrixTensor, [&transMatrix](const long4_16a &c){ return transMatrix[c.y]; });

        benchutils::warmup_and_exec(state, BENCH_WARPAFFINE_WARMUP_ITERATIONS,
            [&op, &src, &dst, &transMatrixTensor, &flags, &borderType, &borderValue](cudaStream_t s) {
                op(s, src, dst, transMatrixTensor, flags, borderType, borderValue);
            });
    }
}
CVCUDA_BENCH_SKIP_ERRORS(state)

// clang-format on

// Use auto-generated type list from bench_params.json

NVBENCH_BENCH_TYPES(warpaffine, NVBENCH_TYPE_AXES(BENCH_WARPAFFINE_TYPES))
BENCH_WARPAFFINE_AXES;
